### ISSUE TITLE
Safari handling of Blobs is broken due to Range workaround.

### ISSUE THREAD

**Issue by thedracle** [brion/stream-file#9]
I am currently hitting the following when using a Blob URL in Safari 11:

```
[Log] [14839.3ms] demuxer ran out of data (on-demand.bundle.js, line 198626)
[Log] [14839.3ms] demuxer loading more data (on-demand.bundle.js, line 198626)
[Log] [14904.2ms] fireEvent framecallback (on-demand.bundle.js, line 198626)
[Log] [14904.3ms] fireEvent timeupdate (on-demand.bundle.js, line 198626)
```

And eventually after it hits:
```
[Log] [15697.6ms] onbufferlow (on-demand.bundle.js, line 198626)
[Log] [15883.1ms] onbufferlow (on-demand.bundle.js, line 198626)
[Log] [16068.7ms] onbufferlow (on-demand.bundle.js, line 198626)
[Log] [16254.5ms] onbufferlow (on-demand.bundle.js, line 198626)
[Log] [16440.2ms] onbufferlow (on-demand.bundle.js, line 198626)
[Log] [16626.1ms] onbufferlow (on-demand.bundle.js, line 198626)
[Log] [16811.8ms] onstarved: halting audio due to starvation (on-demand.bundle.js, line 198626)
[Log] [16811.8ms] pausing at 15.975328798185956 (on-demand.bundle.js, line 198626)
```

Basically it plays a couple seconds just fine, and then locks up.

If I download the blob to disk, and serve it using a webserver, and set the .src to the link on the server it plays back fine.

Maybe something to do with reading the blob too fast? There are probably a lot of differences here, and this doesn't seem to be a normal use case for you.

Also this appears to be being triggered in readBytesAndWait right before failure, with 'buffering' always set to true:
```
function readBytesAndWait() {
	if (stream.buffering || stream.seeking) {
		log('readBytesAndWait during i/o');
		return;
        }
```

**Comment by thedracle** [comment]
So the following recording I just took will reproduce it:

https://s3-us-west-1.amazonaws.com/projectdrop/repro.webm

Basically I count to five three times. Setting src directly will play it back all three times, setting it from a blob via fetch will play it twice.

**Comment by thedracle** [comment]
Hi Brion--- testing with 1.5.1, I'm not sure if there is any way to specify the content-type header via a Blob:

let myBlob = new Blob(myArrayOfWebM, {type: "video/webm"});
or

let myBlob = new Blob(myArrayOfWebM, {'content-type': "video/webm"});
Don't seem to do it.

Neither does specifying 'type' in the options provided to OGVPlayer.

I'm not sure if just from a blob URL it is possible to specify these headers.

**Comment by thedracle** [comment]
So- it seems like the buffering issue I detailed above is something to do with the 'stream' module, and specifically the BinaryStringBackend that is getting selected.

It ends up reading only a portion of the blob via its onXHRProgress callback, and delivering it to the cache, and never delivers any future slices.

It always seems to be around a 1MB chunk, and so the cache actually ends up running out of data.

I added logs at backend/binary-string-backend.js:

  onXHRProgress() {
    const slice = this.xhr.responseText.slice(this.bytesRead);
    if (slice.length > 0) {
      this.bytesRead += slice.length;
      this.emit('buffer', slice);
    }
And at line 151 of stream-file.js:

 } else {
              backend.on('buffer', (buffer) => {
                if (backend === this._backend) {
                  this._cache.write(buffer);
                }
              });
So, for a recording of the size 1419592, I see logged:
[Log] DOWNLOADED SLICE: 1048576 (on-demand.bundle.js, line 205169)
[Log] WRITING BUFFER TO CACHE: 1048576 (on-demand.bundle.js, line 203568)

And this is basically the final amount of data that is available to ogv.js indefinitely.

I wonder if perhaps it's possible for me to use a different backend, or if there is something wrong in my browser (Safari 11) that is causing the XMLHttpRequest to the blob url not provide all of the data.

Thanks so much for your help so far and making this awesome project.

**Comment by thedracle** [comment]
So, it seems this issue is probably completely isolated to Safari, and it has something to do with the below workaround in backend.js on line 124:

```    
         if (this.offset != foundPosition) {
              //
              // Safari sometimes messes up and gives us the wrong chunk.
              // Seems to be a general problem with Safari and cached XHR ranges.
              //
              // Interestingly, it allows you to request _later_ ranges successfully,
              // but when requesting _earlier_ ranges it returns the latest one retrieved.
              // So we only need to update the cache-buster when we rewind and actually
              // get an incorrect range.
              //
              // https://bugs.webkit.org/show_bug.cgi?id=82672
              //
              console.log('Expected start at ' + this.offset + ' but got ' + foundPosition +
                '; working around Safari range caching bug: https://bugs.webkit.org/show_bug.cgi?id=82672');
              this.cachever++;
              this.emit('cachever');
              this.abort();
              oncomplete();
              this.load().then(resolve).catch(reject);
              return;
            }
```

Commenting the above section out makes the entire clip play, and the cache to obtain all of the data from the blob, but probably breaks seek?

I expect the workaround must not work for Blobs for some reason, but Safari is still experiencing the bug where the XHR range start is returning as 0 when it shouldn't.

**Comment by brion** [comment]
Ah, that's a fun interaction. :D Worst case, I can end up doing a workaround for blobs where the whole thing gets loaded from a blob into a giant ArrayBuffer... but I'll see if I can avoid breaking the XHR.

**Comment by brion** [comment]
What seems to be happening is that when it reads beyond a chunk beyond the end of a blob, the 'Range' header specified a range beyond the edge -- Safari seemed to reject that. Fix: tweak the Range on the last chunk to max out at the size of the file (which we already got when we loaded the first chunk successfully) and not go beyond.

Along with a hack to force WebM, this gets your sample running in ogv.js Safari 11 for me. Will make one more change to ogv.js to just detect magic bytes, it's easy to distinguish Ogg from WebM that way...
